lib/spack/spack/test/cmd/dev_build.py

```diff
@@ -215,6 +215,45 @@ def test_dev_build_env(tmpdir, install_mockery, mutable_mock_env_path):
         assert f.read() == spec.package.replacement_string
 
 
+def test_dev_build_env_with_vars(tmpdir, install_mockery, mutable_mock_env_path, monkeypatch):
+    """Test Spack does dev builds for packages in develop section of env (path with variables)."""
+    # setup dev-build-test-install package for dev build
+    build_dir = tmpdir.mkdir("build")
+    spec = spack.spec.Spec(f"dev-build-test-install@0.0.0 dev_path={build_dir}")
+    spec.concretize()
+
+    # store the build path in an environment variable that will be used in the environment
+    monkeypatch.setenv("CUSTOM_BUILD_PATH", build_dir)
+
+    with build_dir.as_cwd():
+        with open(spec.package.filename, "w") as f:
+            f.write(spec.package.original_string)
+
+    # setup environment
+    envdir = tmpdir.mkdir("env")
+    with envdir.as_cwd():
+        with open("spack.yaml", "w") as f:
```

Comment:
same

Reply:
These nested `with` statements have the following structure

```python
with env.as_cwd():
  with open() as f:
    f.write()
  env("create", ...) # from f
  with ev.read(...): # created by previous line
    install()
```

switching to

```python
with env, open() as f, ev.read():
  f.write()
  env()
  install()
```

is equivalent to

```python
with env:
  with open() as f:
    with ev.read(): # environment does not exist yet
      f.write()
      env("create") # is going to read the file `f` while it is still open for writing
      install()
```